src/protocol/context/mod.rs

```diff
@@ -0,0 +1,94 @@
+use crate::ff::Field;
+use crate::helpers::messaging::{Gateway, Mesh};
+use crate::helpers::Role;
+use crate::protocol::mul::SecureMul;
+use crate::protocol::prss::{
+    Endpoint as PrssEndpoint, IndexedSharedRandomness, SequentialSharedRandomness,
+};
+use crate::protocol::reveal::Reveal;
+use crate::protocol::{Step, Substep};
+use crate::secret_sharing::SecretSharing;
+use std::sync::Arc;
+
+mod malicious;
+mod semi_honest;
+
+pub use malicious::MaliciousProtocolContext;
+pub use semi_honest::SemiHonestProtocolContext;
+
+/// Context used by each helper to perform secure computation. Provides access to shared randomness
+/// generator and communication channel.
+pub trait ProtocolContext<F: Field>:
```

Comment:
no more `bind`

Reply:
At the start of the meeting yesterday, @martinthomson and I were discussing checking for bound contexts using the type system. https://github.com/private-attribution/ipa/compare/main...andyleiserson:ipa:context2 has some changes I was working on in that direction. I was thinking it might make sense to get rid of the separate `record_id` parameter and just pass it via the context.

I had also worked on some changes to enforce one-time use restrictions. Those are in https://github.com/andyleiserson/ipa/compare/context2...andyleiserson:ipa:context3.

Reply:
Hmm, with this change we don't need `record_id` inside the context, unless we want to change the way how we multiply shares. Currently it is  `ctx.multiply(record_id, a, b)`, do we want `ctx.multiply(a, b)`?

Reply:
I kinda like what Andy has done here.  The `ctx.multiply(a, b)` would be `ctx.bind(record).multiply(a, b)`, which seems like a solid approach, though it looks like `ctx.multiply(record, a, b)` also works fine.

into_parts() taking a reference seems unfortunate; it would be good to understand what parts of that don't work; is it that these functions also narrow()?  It seems like you might want something on which you can narrow only in addition to the usable bits if that is the case.

The ergonomics of into_parts leaves a little to be desired.  It's not awful, but it loses some of the simplicity.

Reply:
> into_parts() taking a reference seems unfortunate

Agreed, this was temporary so that I didn't have to update all references to the context all at once. I would want to fix it before merging anything.

>  It seems like you might want something on which you can narrow only in addition to the usable bits if that is the case.

Also agreed, this was on my to do list.

Reply:
I am a little hesitant because `ctx.multiply(a, b)` syntax invites us to solve the problem "how to bind once", while `ctx.multiply(record, a, b)` does not.

Reply:
By "how to bind once", do you mean keeping track of whether the context is already bound and issuing a warning if it is, similar to the "context narrowed more than once" warning that is removed by this PR?

My proposal is to keep track of that with the type system. So if it compiles, the context is not bound more than once. The benefit of this is there's less chance of a bug due to passing a wrong or inconsistent record_id somewhere.

Reply:
I guess I am debating whether we need compile-time check vs the runtime check that already exists. There are some really nice things we can do with bounded contexts - for example implement multiplication and reveal for contexts that carry `record_id` , so nobody will be able to call `multiply` on `Context<.., AnyRecord>`. On the other hand, it requires some machinery and makes code a bit harder to understand? Let me add this topic to the agenda for the Thursday sync-up - there are some really nice things we can do with the model you proposed.